File: mergesort.c
/*
 coder : bobby
 channel : Believer 01
 algorithm : Merge sort
 date_coded: 08/06/2020
 time complexity : O(n*log(n))
*/
#include<stdio.h>
#include<stdlib.h>

void merge(int *arr,int * l_arr,int * r_arr,int l_size,int r_size,int n)
{
    int i=0,j=0,k=0;
    while(i<l_size && j<r_size)
    {
        if(l_arr[i]<r_arr[j])
        {
            arr[k]=l_arr[i];
            i++;
        }
        else
        {
            arr[k]=r_arr[j];
            j++;
        }
        k++;
    }
    while(i<l_size)
    {
        arr[k]=l_arr[i];
        i++;
        k++;
    }
    while(j<r_size)
    {
        arr[k]=r_arr[j];
        j++;
        k++;
    }
}
void merge_sort(int *arr,int n)
{
    if(n>1)
    {
        int l_size=n/2,r_size=n-l_size;
        int *l_arr=(int*)malloc(sizeof(int*)*l_size);
        int *r_arr=(int*)malloc(sizeof(int*)*r_size);
        int i=0,j=0;
        for(i =0;i<l_size;i++)
        {
            l_arr[i]=arr[j];
            j++;
        }
        for(i=0;i<r_size;i++)
        {
            r_arr[i]=arr[j];
            j++;
        }
        merge_sort(l_arr,l_size);
        merge_sort(r_arr,r_size);
        merge(arr,l_arr,r_arr,l_size,r_size,n);
    }
}

void print_array(char *message,int *arr,int n)
{
    printf("%s : ",message);
    int i=0;
    for(i=0;i<n;i++)
        printf("%d ",arr[i]);
    printf("\n");
}
int main()
{
    int arr[]={4,8,1,7,2,6,5,3};
    int n=sizeof(arr)/sizeof(arr[0]);
    print_array("Before Sorting",arr,n);
    merge_sort(arr,n);
    print_array("After Sorting",arr,n);
    return 0;
}

File: merge_sort.py
# coder: bobby
# channel: Believer 01 
# date_coded: 08/06/2020 
# Merge Sort
# Time Complexity: O(n*log(n))
# 4 8 1 7 2 6 5 3

def merge(arr,first_array,second_array,first_part,second_part,n):
    i=0
    j=0
    k=0
    while i<first_part and j<second_part:
        if first_array[i]<second_array[j]:
            arr[k]=first_array[i]
            i+=1
        else:
            arr[k]=second_array[j]
            j+=1
        k+=1
    while i<first_part:
        arr[k]=first_array[i]
        i+=1
        k+=1
    while j<second_part:
        arr[k]=second_array[j]
        j+=1
        k+=1
def merge_sort(arr,n):
    if n>1:
        first_part=n//2
        second_part=n-first_part
        first_array=[]
        second_array=[]
        j=0
        for i in range(0,first_part):
            first_array.append(arr[j])
            j+=1
        for i in range(0,second_part):
            second_array.append(arr[j])
            j+=1
        merge_sort(first_array,first_part)
        merge_sort(second_array,second_part)
        merge(arr,first_array,second_array,first_part,second_part,n)
def print_array(message,arr,n):
    print(message,end=" : ")
    for i in range(0,n):
        print(arr[i],end=" ")
    print()
def main():
    arr=[4,8,1,7,2,6,5,3]
    n=len(arr)
    print_array('Before Sorting',arr,n)
    merge_sort(arr,n)
    print_array('After Sorting',arr,n)
main()
